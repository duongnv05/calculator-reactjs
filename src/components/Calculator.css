._calculator {
  display: flex;
  flex-direction: column;
  max-width: 295px;
  height: fit-content;
  margin-top: 13px;
}

._calculator .screen-control {
  height: 80px;
  background-color: var(--main-color);
  text-align: right;
  color: white;
  font-weight: 600;
  padding: 10px;
  display: flex;
  flex-direction: column;
  border-top-left-radius: 5px;
  border-top-right-radius: 5px;
}

._calculator .screen-control .result {
  height: 70%;
  font-size: 30px;
  overflow: hidden;
  display: flex;
  justify-content: flex-end;
}

._calculator .screen-control .content-calculator {
  height: 30%;
  font-size: 20px;
  overflow: hidden;
  display: flex;
  justify-content: flex-end;
}

/* BUTTON CONTROL */
._calculator .button-control .wrapper {
  display: flex;
  flex-wrap: wrap;
  width: 100%;
  padding-bottom: 15px;
  background: white;
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
}

._calculator .button-control .btn-control {
  width: 25%;
  background-color: white;
  padding: 28px;
  font-size: 15px;
  font-weight: 500;
}
._calculator .button-control .btn-control:hover {
  z-index: 1;
  border-color: rgba(0,0,0,.12);
  /* box-shadow: 0 0 0.8125rem 0 rgba(0,0,0,.05); */
  box-shadow: 0 9px #999;
  border-bottom-left-radius: 5px;
  border-bottom-right-radius: 5px;
}
._calculator .button-control .btn-control:active {
  /* background-color: #3e8e41; */
  box-shadow: 0 5px #666;
  transform: translateY(4px);
}
._calculator .button-control .btn-control.merce-3-row {
  width: 75%;
}