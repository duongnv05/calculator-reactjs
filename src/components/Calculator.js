import React, { Component } from 'react';

import './Calculator.css';

export default class Calculator extends Component {
  constructor() {
    super();
    this.state = {
      content: '0',
      result: '0'
    };

    this.onInputCalculator = this.onInputCalculator.bind(this);
    this.onClearButton = this.onClearButton.bind(this);

    this.onTouchStartClear = this.onTouchStartClear.bind(this);
    this.onTouchEndClear = this.onTouchEndClear.bind(this);
    this.onHandleKeyBoard = this.onHandleKeyBoard.bind(this);
  }

  componentDidMount() {
    document.addEventListener('keydown', this.onHandleKeyBoard);
  }

  componentWillUnmount() {
    document.removeEventListener('keydown', this.onHandleKeyBoard);
  }

  onHandleKeyBoard(evt) {
    switch(evt.key) {
      case '+':
      case '-':
      case '/':
      case '=':
      case '1':
      case '2':
      case '3':
      case '4':
      case '5':
      case '6':
      case '7':
      case '8':
      case '9':
        this.onInputCalculator(evt.key);
        break;
      case 'Backspace':
        this.onInputCalculator('clear');
        break;
      case 'Enter':
        this.onInputCalculator('=');
        break;
    }
  }

  onTouchStartClear() {
    this.btnClearTimer = setTimeout(() => {
      this.setState({
        content: '0',
        result: '0'
      });
    }, 800)

    this.onInputCalculator('clear');
  }

  onTouchEndClear() {
    clearTimeout(this.btnClearTimer);
  }

  onClearButton() {
    if(this.state.content == '0') return;

    let temp = this.state.content.slice(0, this.state.content.length - 1);
    this.setState({
      content: (temp == '' ? '0' : temp)
    })

    // this.onInputCalculator('clear');
  }

  onInputCalculator(val) {
    let _result = '0';
    let preState = this.state.content;
    let futureState;

    if(preState == '0') { preState = '' }

    if(val == 'clear') {
      futureState = preState.slice(0, preState.length - 1);
    } else {
      futureState = preState + val;
    }

    if(futureState == '' 
    //- prevent input '/' first
    || futureState == '/') { preState = '0'; futureState = '0' }

    //- calculate result
    let reResult = RegExp(/^\s*([-+]?)(\d+)$|^\s*([-+]?)(\d+)(?:\s*([-+*\/])\s*((?:\s[-+])?\d+)\s*)+$/, 'g');
    if(reResult.test(futureState)) {
      _result = eval(futureState);
    } else if(reResult.test(preState)) {
      _result = eval(preState);
    } else {
      _result = eval(preState.slice(0, preState.length - 1))
    }

    if(val == '=') {
      //- handle calculating string content
      this.setState({
        // content: _result,
        result: _result ? _result : '0'
      });
      return;
    }

    //- the input is valid
    // let re = RegExp(/^([-+]?)(\d*)([-+*\/]?$)|(^\s*([-+]?)(\d+)(?:\s*([-+*\/])\s*((?:\s[-+])?\d+)\s*(?:[-+*\/]?))+$)/, 'g');
    let re = RegExp(/((^([-,+]))(\d+)(([-+*\/]?$))$)|(^(([-]?)(\d*))$|(^(\d{0,})+([-+*\/]?$)$)$)|(^\s*([-+]?)(\d+)(?:\s*([-+*\/])\s*((?:\s[-+])?\d+)\s*(?:[-+*\/]?))+$)/, 'g');

    if(!re.test(futureState)) return;

    this.setState({
      content: futureState,
      result: _result
    });
  }

  render() {
    const { content, result } = this.state;
    return (
      <div className="_calculator">
        {/* screen control */}
        <div className="screen-control">
          {/* result control */}
          <div className="result" title={result}>
            { result }
          </div>
          <div className="content-calculator">
            { content }
          </div>
        </div>
        {/* button controls */}
        <div className="button-control">
          <div className="wrapper">
            <button className="btn-control merce-3-row clear" 
              onTouchStart={this.onTouchStartClear} 
              onTouchEnd={this.onTouchEndClear} 
              onMouseDown={this.onTouchStartClear} 
              onMouseUp={this.onTouchEndClear} 
              onMouseLeave={this.onTouchEndClear}
            >
              Clear
            </button>
            <button className="btn-control /" onClick={()=>this.onInputCalculator('/')}>/</button>
            <button className="btn-control 7" onClick={()=>this.onInputCalculator('7')}>7</button>
            <button className="btn-control 8" onClick={()=>this.onInputCalculator('8')}>8</button>
            <button className="btn-control 9" onClick={()=>this.onInputCalculator('9')}>9</button>
            <button className="btn-control -" onClick={()=>this.onInputCalculator('-')}>-</button>
            <button className="btn-control 4" onClick={()=>this.onInputCalculator('4')}>4</button>
            <button className="btn-control 5" onClick={()=>this.onInputCalculator('5')}>5</button>
            <button className="btn-control 6" onClick={()=>this.onInputCalculator('6')}>6</button>
            <button className="btn-control +" onClick={()=>this.onInputCalculator('+')}>+</button>
            <button className="btn-control 1" onClick={()=>this.onInputCalculator('1')}>1</button>
            <button className="btn-control 2" onClick={()=>this.onInputCalculator('2')}>2</button>
            <button className="btn-control 3" onClick={()=>this.onInputCalculator('3')}>3</button>
            <button className="btn-control =" onClick={()=>this.onInputCalculator('=')}>=</button>
          </div>
        </div>
      </div>
    )
  }
}