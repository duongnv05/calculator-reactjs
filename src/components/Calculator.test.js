import React from 'react';
import Calculator from './Calculator';
import { shallow } from 'enzyme';
describe('<Calculator />', () => {
  it('get result when press(or click) =', () => {
    const wrapper = shallow(<Calculator />);

    wrapper.setState({
      content: '1+2+3/3+5'
    })

    wrapper.find('button').at(13).simulate('click');

    expect(wrapper.state('result')).toEqual(9);
  })

  it('get result when press(or click) = (negative number)', () => {
    const wrapper = shallow(<Calculator />);

    wrapper.setState({
      content: '-11+2+3/3+5'
    })

    wrapper.find('button').at(13).simulate('click');

    expect(wrapper.state('result')).toEqual(-3);
  })

  it('user input invalid', () => {
    const wrapper = shallow(<Calculator />);

    wrapper.setState({
      content: '1+2+3/3+5'
    })

    wrapper.find('button').at(1).simulate('click');

    expect(wrapper.state('content')).toEqual('1+2+3/3+5/');
    expect(wrapper.state('result')).toEqual(9);
  })

  it('press /, invalid input the first time', () => {
    const wrapper = shallow(<Calculator />);

    wrapper.setState({
      content: '0'
    })

    wrapper.find('button').at(1).simulate('click');

    expect(wrapper.state('content')).toEqual('0');
    expect(wrapper.state('result')).toEqual(0);
  })
})