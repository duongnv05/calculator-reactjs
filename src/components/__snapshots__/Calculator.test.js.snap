// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Calcular changes the result when user input value 1`] = `
<div
  className="_calculator"
>
  <div
    className="screen-control"
  >
    <div
      className="result"
    >
      0
    </div>
    <div
      className="content-calculator"
    >
      0
    </div>
  </div>
  <div
    className="button-control"
  >
    <div
      className="wrapper"
    >
      <button
        className="btn-control merce-3-row"
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
      >
        Clear
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        /
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        7
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        8
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        9
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        -
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        4
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        5
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        6
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        +
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        1
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        2
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        3
      </button>
      <button
        className="btn-control"
        onClick={[Function]}
      >
        =
      </button>
    </div>
  </div>
</div>
`;
